From 72c1510c75137a894b38aacb1994c6b18e0c9796 Mon Sep 17 00:00:00 2001
From: "po.cheng" <po.cheng@adlinktech.com>
Date: Thu, 23 Jan 2025 16:59:44 +0800
Subject: [PATCH 1/3] Add Falcon Mode support for iMX8M 1of3 [PATCH] From
 b18d1bf9304d0382de414ff2b64ec9190b17aebd Mon Sep 17 00:00:00 2001 From: Elena
 Popa <elena.popa@nxp.com> Date: Wed, 11 Sep 2024 18:30:38 +0300 Subject:
 [PATCH] Add Falcon Mode support for iMX8M and iMX9

Signed-off-by: po.cheng <po.cheng@adlinktech.com>
---
 common/spl/spl.c     | 26 ++++++++++++++++++++++++++
 common/spl/spl_fat.c |  2 +-
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/common/spl/spl.c b/common/spl/spl.c
index b452d4feeb2f..f147878120dd 100644
--- a/common/spl/spl.c
+++ b/common/spl/spl.c
@@ -36,6 +36,8 @@
 #include <fdt_support.h>
 #include <bootcount.h>
 #include <wdt.h>
+#include <asm/mach-imx/gpio.h>
+#include <linux/delay.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -47,6 +49,14 @@ DECLARE_GLOBAL_DATA_PTR;
 #define CONFIG_SYS_MONITOR_LEN	(200 * 1024)
 #endif
 
+#ifdef CONFIG_ARCH_IMX8M
+#define RESET_ETH0_GPIO IMX_GPIO_NR(4,22)
+#ifdef CONFIG_IMX8MP
+#define RESET_ETH1_GPIO IMX_GPIO_NR(4,2)
+#endif
+#define USDHC_GPIO_PAD_CTRL (PAD_CTL_HYS | PAD_CTL_DSE1)
+#endif
+
 u32 *boot_params_ptr = NULL;
 
 #if CONFIG_IS_ENABLED(BINMAN_SYMBOLS)
@@ -743,6 +753,22 @@ void board_init_r(gd_t *dummy1, ulong dummy2)
 		}
 	}
 
+#ifdef CONFIG_ARCH_IMX8M
+	/*reset eth*/
+	gpio_request(RESET_ETH0_GPIO, "reset_eth0_gpio");
+	gpio_direction_output(RESET_ETH0_GPIO, 0);
+	mdelay(1);
+	gpio_direction_output(RESET_ETH0_GPIO, 1);
+	mdelay(1);
+#ifdef CONFIG_IMX8MP
+	gpio_request(RESET_ETH1_GPIO, "reset_eth1_gpio");
+	gpio_direction_output(RESET_ETH1_GPIO, 0);
+	mdelay(1);
+	gpio_direction_output(RESET_ETH1_GPIO, 1);
+	mdelay(1);
+#endif
+#endif
+
 #if CONFIG_IS_ENABLED(BOARD_INIT)
 	spl_board_init();
 #endif
diff --git a/common/spl/spl_fat.c b/common/spl/spl_fat.c
index 5b270541fcc7..6e7788840624 100644
--- a/common/spl/spl_fat.c
+++ b/common/spl/spl_fat.c
@@ -40,7 +40,7 @@ static int spl_register_fat_device(struct blk_desc *block_dev, int partition)
 	return err;
 }
 
-static ulong spl_fit_read(struct spl_load_info *load, ulong file_offset,
+ulong __weak spl_fit_read(struct spl_load_info *load, ulong file_offset,
 			  ulong size, void *buf)
 {
 	loff_t actread;
-- 
2.25.1

